### PR TITLE
Move handling of empty from obscure function to main email handler class 

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
[Move handling of empty from obscure function to main email handler class](https://github.com/civicrm/civicrm-core/commit/5384c03efa2324543dd331a20aa2dc580c3ce455)

Before
----------------------------------------
The handling for https://issues.civicrm.org/jira/browse/CRM-19215 is untested & in a weird place

After
----------------------------------------
 I had hoped when adding a test it would now be obsolete. However, the test showed it is still needed for that situation - so I moved it to a more sensible location. Note that I expect the function I moved it out of to be entirely removed 

Technical Details
----------------------------------------

Comments
----------------------------------------

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Utils/Mail/EmailProcessor.php [hunk 1]
```diff
@@ -117,13 +117,15 @@ private static function _process($civiMail, $dao, $is_create_activities) {
 
           // if its the activities that needs to be processed ..
           try {
-            $mailParams = CRM_Utils_Mail_Incoming::parseMailingObject($mail, $createContact, FALSE, $emailFields);
+            $mailParams = CRM_Utils_Mail_Incoming::parseMailingObject($mail, $createContact, [$incomingMail->getFrom()], $emailFields);
```
Comment (eileenmcnaughton) [inline CRM/Utils/Mail/EmailProcessor.php:120]: this is a fugly way to pass it in - but this function doesn't have long-to-live in general.... or not as it currently appears

**Comment by eileenmcnaughton** [comment]
I just rebased this over https://github.com/civicrm/civicrm-core/pull/27356 as that is the higher priority change & I didn't want this to make that stale!

File: CRM/Utils/Mail/EmailProcessor.php [hunk 2]
```diff
@@ -77,12 +77,13 @@ public static function processActivities() {
...
-      $emailFields = array_filter(array_unique(array_merge(explode(",", $dao->activity_targets), explode(",", $dao->activity_assignees), explode(",", $dao->activity_source))));
-      $createContact = !($dao->is_contact_creation_disabled_if_no_match ?? FALSE);
-    }
+    $isBounceProcessing = $dao->is_default;
+    $targetFields = array_filter(explode(',', $dao->activity_targets));
+    $assigneeFields = array_filter(explode(",", $dao->activity_assignees));
+    $sourceFields = array_filter(explode(",", $dao->activity_source));
+    // create an array of all of to, from, cc, bcc that are in use for this Mail Account, so we don't create contacts for emails we aren't adding to the activity.
+    $emailFields = array_merge($targetFields, $assigneeFields, $sourceFields);
+    $createContact = !($dao->is_contact_creation_disabled_if_no_match);
```
Comment (larssandergreen) [inline CRM/Utils/Mail/EmailProcessor.php:86]: Maybe move 80-86 down into the try block at 120, just before it is used? Might make more sense (than this weirdly placed code that I added recently).

File: CRM/Utils/Mail/EmailProcessor.php [hunk 3]
```diff
@@ -117,34 +118,38 @@ private static function _process($civiMail, $dao, $is_create_activities) {
...
+            }
+            else {
+              $activityParams['source_contact_id'] = $mailParams[$dao->activity_source][0]['id'];
 
-            $activityParams['source_contact_id'] = $mailParams[$dao->activity_source][0]['id'];
-
-            $activityContacts = ['target_contact_id' => 'activity_targets', 'assignee_contact_id' => 'activity_assignees'];
-            foreach ($activityContacts as $activityContact => $daoName) {
-              $activityParams[$activityContact] = [];
-              $activityKeys = array_filter(explode(",", $dao->$daoName));
```
Comment (larssandergreen) [inline CRM/Utils/Mail/EmailProcessor.php:132]: Lost this line

File: CRM/Utils/Mail/EmailProcessor.php (hunk 2 above)
Comment (eileenmcnaughton) [reply CRM/Utils/Mail/EmailProcessor.php:86]: I think I was gonna do that & then I thought that would move it into the look & maybe not do it every row

File: CRM/Utils/Mail/EmailProcessor.php (hunk 3 above)
Comment (eileenmcnaughton) [reply CRM/Utils/Mail/EmailProcessor.php:132]: @larssandergreen so `  $activityKeys = array_filter(explode(",", $dao->$daoName));` was from looking up the dao again but now it comes from the earlier lookup 

```
   $activityContacts = [
                'target_contact_id' => $targetFields,
                'assignee_contact_id' => $assigneeFields,
              ];
```

File: CRM/Utils/Mail/EmailProcessor.php (hunk 2 above)
Comment (larssandergreen) [reply CRM/Utils/Mail/EmailProcessor.php:86]: Makes sense - that's why I put it up there!

File: CRM/Utils/Mail/EmailProcessor.php (hunk 3 above)
Comment (larssandergreen) [reply CRM/Utils/Mail/EmailProcessor.php:132]: Right, makes sense

**Comment by larssandergreen** [comment]
Does it make sense to r-run this now or better with the whole thing at once?

**Comment by eileenmcnaughton** [comment]
@larssandergreen I think that if I can get this merged into the rc then I'll stop on this for now cos this has the actual fixes in it & I don't want to keep cleaning up in master cos it would invalidate people's efforts to test

**Comment by eileenmcnaughton** [comment]
@larssandergreen Although I do have a few last thoughts on cleanup .... mostly move the attachments handling to the Incoming class & de-nest the contacts array - so if you did want to merge this now I could put up a couple more cleanups against the rc & then testing could be on the rc as a whole

**Comment by eileenmcnaughton** [comment]
@larssandergreen now I'm confused between the 2 PRs - I propose that I push the patches from this into the other one - which has the better PR template and then we merge it & keep testing the full rc.

I just deployed this into production so that would align the rc with what we have deployed & are having good results with. However, per my comments on the other one I think a couple more fixes will improve usability

**Comment by eileenmcnaughton** [comment]
As discussed consolidating to https://github.com/civicrm/civicrm-core/pull/27356

**Pull request closed without merging**